### PR TITLE
[wit-component] add shared-everything linking support

### PR DESCRIPTION
**Pull request by dicej**
Per https://github.com/WebAssembly/component-model/blob/main/design/mvp/examples/SharedEverythingDynamicLinking.md and https://hackmd.io/IlY4lICRRNy9wQbNLdb2Wg.

This adds a new `component link` subcommand, which resembles `component new` but accepts an arbitrary number of "dynamic library" modules (as defined by https://github.com/WebAssembly/tool-conventions/blob/main/DynamicLinking.md), linking them together to create a component whose type is the union of the component types found in the input modules (via the usual custom section representation).  This process can also be invoked programmatically via a new `Linker` API.

`Linker` analyzes and topologically sorts the input modules, then synthesizes two additional modules:

- `main`: hosts the component's single memory and function table and exports any functions needed to break dependency cycles discovered in the input modules. Those functions use `call.indirect` to invoke the real functions, references to which are placed in the table by the `init` module.

- `init`: populates the function table as described above, initializes global variables per the dynamic linking tool convention, and calls any static constructors and/or link-time fixup functions

Finally, `Linker` generates a directed graph of modules and their instantiations and passes it to `ComponentEncoder` to actually produce the component.

The implementation relies on a few modifications to the existing code in `ComponentEncoder` and friends.  Specifically, I've expanded the meaning of "adapter" to include dynamic libraries as well as traditional adapters.  The former are distiguished from the latter by the presence of a new `LibraryInfo` struct.  This approach minimized the number of code changes in the interest of keeping this PR comprehensible, but we will probably want to refactor `ComponentEncoder` in the future to make the distinction between traditional adapters and dynamic libraries clearer, possibly generalizing how we analyze and instantiate modules of various flavors.

I've created a demo which ties everything together, for reference: https://github.com/dicej/component-linking-demo.  It includes `wasi-sdk` and `wasi-libc` patches which have not yet been upstreamed; I'm planning to work on that next.

### CODE REVIEW

**Comment by dicej** [comment]
I need to add some more tests before this will be ready to merge, but I'm opening a draft PR in case anyone wants to comment in the meantime.

**Comment by dicej** [comment]
TIL (most of) the `wasm-tools` tests need to be able to run in a Wasm guest, and I broke that by using `wasmtime` in the linking test.  I guess I have two options to address that: either rewrite the linking test to not use `wasmtime` (i.e. turn it into one of those "matches expected output WAT" tests) or exclude the linking test from the set of tests run inside Wasm.  Anyone have a preference?

File: crates/wit-component/tests/linking.rs
Comment (alexcrichton) [inline crates/wit-component/tests/linking.rs:11]: One comment I'd have on this is that I think it'd be best if the test suite here could be structured like the other wit-component tests where there's a folder-of-tests and each test is a set of files configuring what needs to be done. The test-runner is more-or-less "go use the library equivalents of `wasm-tools` CLI commands to execute tests".

I've opened https://github.com/bytecodealliance/wasm-tools/pull/1135 which may be able to assist with that, but I've found this style of tests is beneficial for reading tests, writing tests, and debugging tests. (e.g. basically making it such that it's possible to play with everything on the CLI rather than requiring custom Rust code to drive everything.

Comment (dicej) [reply crates/wit-component/tests/linking.rs:11]: I've merged #1135 into https://github.com/dicej/wasm-tools/pull/new/linking-wat and added linking support to the components.rs test runner.  When 1135 is merged, I 'll rebase this PR and pull that in.  Meanwhile, I'll continue adding tests.

BTW, is it okay if I keep the one end-to-end test in linking.rs in addition to all the WAT tests?  It would be nice to have at least one test that actually runs the component.

Comment (dicej) [reply crates/wit-component/tests/linking.rs:11]: I've finished adding tests to the `linking-wat` branch of my fork, so just waiting on 1135 to land at this point.

File: crates/wit-component/tests/linking.rs
Comment (alexcrichton) [inline crates/wit-component/tests/linking.rs:139]: Could this perhaps be replaced with the `@dylink` text format in the above `*.wat` files?

File: crates/wit-component/src/linking.rs
Comment (alexcrichton) [reply crates/wit-component/src/linking.rs:268]: I see now this is used quite a bit throughout this module, so perhaps something like:

```rust
fn const_u32(a: u32) -> ConstExpr { .. }
```

or something like that?

File: crates/wit-component/src/linking.rs
Comment (alexcrichton) [inline crates/wit-component/src/linking.rs:751]: Could this assert that the insertion didn't replace anything?

File: crates/wit-component/src/linking.rs
Comment (alexcrichton) [inline crates/wit-component/src/linking.rs:837]: One thing to be careful of here is that `deps` is a `HashSet` which has non-deterministic iteration order between runs. This might be a good place to use `IndexSet` to ensure the output of the `link` command is deterministic.

File: crates/wit-component/src/linking.rs
Comment (alexcrichton) [inline crates/wit-component/src/linking.rs:827]: I'm a bit curious about how this works in the presence of cycles. I've never worked with topo sorts with cycles before, so naively I'd interpret the checks below as they handle a cycle of A->B->A but wouldn't handle deeper cycles and may recurse infinitely? I haven't worked with this before so I'm not sure though.

Put another way I suppose, what's the meaning of a topo sort in the presence of cycles?

File: crates/wit-component/src/linking.rs
Comment (alexcrichton) [inline crates/wit-component/src/linking.rs:874]: Could this, and insertions below, assert that they all succeed? (e.g. don't replace anything)

File: crates/wit-component/src/linking.rs
Comment (alexcrichton) [inline crates/wit-component/src/linking.rs:901]: Could this have a comment to explain what's going on? It looks like a fixpoint-style loop but I'm not quite sure why that's needed here

Comment (alexcrichton) [reply crates/wit-component/src/linking.rs:901]: I'll also note that there's more `HashMap` iteration here which may run a risk of leading to non-determinism in the output

File: crates/wit-component/tests/components/error-link-duplicate-symbols/error.txt
Comment (alexcrichton) [inline crates/wit-component/tests/components/error-link-duplicate-symbols/error.txt:1]: Could this, and some of the other errors here, perhaps be formatted without `Debug` output?

**Review by alexcrichton — APPROVED** [review]
Looks reasonable to me, thanks for putting in the work on this! This definitely turned out pretty gnarly, but that's to be expected in this problem domain.

I don't intimiately understand the details of dynamic linking but everything appeared pretty reasonable to me here so I'm leaning on you mostly to ensure it's handled alright (and sounds like with the size of the demos you're working with that'll be ok). The main two meta-level comments I have which are also interspersed in some review are:

* It may be worth watching out for hash maps and iteration order leading to nondeterministic output. In my opinion maps are fine when it's known they won't affect the output order (e.g. only used for lookup), but maps that are iterated over may want to be `Index{Map,Set}` instead.
* In general I try to assert that insertion into a map or set succeeds without replacing something to catch possible bugs where I got something wrong by accident. It may be worth adding asserts to all the various places to handle that?

File: crates/wit-component/src/linking.rs
Comment (dicej) [reply crates/wit-component/src/linking.rs:320]: LLVM (i.e. wasm-ld) expects them to be.  For example:

```
 $ wasm-tools print /opt/wasi-sdk/share/wasi-sysroot/lib/wasm32-wasi/libc.so | grep '"__heap'
  (import "GOT.mem" "__heap_base" (global $__heap_base (;3;) (mut i32)))
  (import "GOT.mem" "__heap_end" (global $__heap_end (;4;) (mut i32)))
```

File: crates/wit-component/src/linking.rs
Comment (dicej) [reply crates/wit-component/src/linking.rs:310]: From https://github.com/WebAssembly/tool-conventions/blob/main/DynamicLinking.md#imports:
> The imported global must be mutable as the dynamic linker might need to modify the value after instantiation.

File: crates/wit-component/src/linking.rs
Comment (alexcrichton) [reply crates/wit-component/src/linking.rs:320]: alas!

File: crates/wit-component/src/linking.rs
Comment (dicej) [reply crates/wit-component/src/linking.rs:827]: In this case, "topo sort in the presence of cycles" means the algorithm makes no guarantees about the relative order of items which participate in a cycle.  If we wanted to, we could refine that a bit so that if A needs 20 things from B and B only needs 3 things from A, then we'd put B first since it doesn't depend on A "as much" as A depends on B, but that was more than I was ready to tackle.

It handles deeper cycles by expecting that `dependencies` includes transitive dependencies, not just the first level of dependencies.

File: crates/wit-component/src/linking.rs
Comment (dicej) [reply crates/wit-component/src/linking.rs:112]: `linking::metadata::Type` only has two variants: `Function` and `Global`, so there's nothing to assert for now.  I'll change this to a `match`, though, which will give us an exhaustiveness check in case we add more variants later.

File: crates/wit-component/src/linking.rs
Comment (dicej) [reply crates/wit-component/src/linking.rs:98]: https://github.com/dicej/wasi-libc/blob/76c7e1e1cfdad577ecd7f61c67ead7a38d62a7c4/libc-top-half/musl/src/misc/dl.c will read the tables we generate here as C structs, so the fields need to be aligned accordingly.

File: crates/wit-component/src/linking.rs
Comment (dicej) [reply crates/wit-component/src/linking.rs:299]: Not necessary, but I find that limiting the scope of variables helps readability by making it clear the value won't be used elsewhere.

File: crates/wit-component/src/linking.rs
Comment (dicej) [reply crates/wit-component/src/linking.rs:820]: In this case, we actually expect it may replace something since multiple libraries may import the same symbol.  I've added a comment to note that.  I've also added assertions to a few places where we don't expect replacement.

File: crates/wit-component/src/linking.rs
Comment (dicej) [reply crates/wit-component/src/linking.rs:908]: I've changed this to an `IndexSet`.

File: crates/wit-component/src/linking.rs
Comment (dicej) [reply crates/wit-component/src/linking.rs:978]: I've added a comment and changed it to an `IndexMap`.

**Comment by dicej** [comment]
Thanks, as always, for the thorough review, @alexcrichton.  I believe I've addressed all your feedback so far.

**Comment by alexcrichton** [comment]
👍 Thanks again!

**Pull request merged**
